source/extensions/filters/network/thrift_proxy/filters/ratelimit/ratelimit.cc

```diff
@@ -0,0 +1,126 @@
+             ? ThriftProxy::FilterStatus::StopIteration
+             : ThriftProxy::FilterStatus::Continue;
+}
+
+void Filter::initiateCall(const ThriftProxy::MessageMetadata& metadata) {
+  ThriftProxy::Router::RouteConstSharedPtr route = callbacks_->route();
+  if (!route || !route->routeEntry()) {
+    return;
+  }
+
+  const ThriftProxy::Router::RouteEntry* route_entry = route->routeEntry();
+  Upstream::ThreadLocalCluster* cluster = config_->cm().get(route_entry->clusterName());
+  if (!cluster) {
+    return;
+  }
+  cluster_ = cluster->info();
+
+  std::vector<RateLimit::Descriptor> descriptors;
+
+  // Get all applicable rate limit policy entries for the route.
+  populateRateLimitDescriptors(route_entry->rateLimitPolicy(), descriptors, route_entry, metadata);
+
+  if (!descriptors.empty()) {
+    state_ = State::Calling;
+    initiating_call_ = true;
+    client_->limit(*this, config_->domain(), descriptors, Tracing::NullSpan::instance());
+    initiating_call_ = false;
+  }
+}
+
+void Filter::onDestroy() {
+  if (state_ == State::Calling) {
+    state_ = State::Complete;
+    client_->cancel();
+  }
+}
+
+void Filter::complete(RateLimit::LimitStatus status, Http::HeaderMapPtr&& headers) {
+  // TODO(zuercher): Store headers to append to a response. Adding them to a local reply (over
+  // limit or error) is a matter of modifying the callbacks to allow it. Adding them to an upstream
+  // response requires either response (aka encoder) filters or some other mechanism.
+  UNREFERENCED_PARAMETER(headers);
+
+  state_ = State::Complete;
+
+  switch (status) {
+  case RateLimit::LimitStatus::OK:
+    cluster_->statsScope().counter("ratelimit.ok").inc();
+    break;
+  case RateLimit::LimitStatus::Error:
+    cluster_->statsScope().counter("ratelimit.error").inc();
+    break;
+  case RateLimit::LimitStatus::OverLimit:
+    cluster_->statsScope().counter("ratelimit.over_limit").inc();
+    break;
+  }
+
+  if (status == RateLimit::LimitStatus::OverLimit &&
+      config_->runtime().snapshot().featureEnabled("ratelimit.thrift_filter_enforcing", 100)) {
+    state_ = State::Responded;
+    callbacks_->sendLocalReply(
+        ThriftProxy::AppException(ThriftProxy::AppExceptionType::InternalError, "over limit"),
+        false);
+    callbacks_->streamInfo().setResponseFlag(StreamInfo::ResponseFlag::RateLimited);
+  } else if (status == RateLimit::LimitStatus::Error) {
+    if (config_->failureModeAllow()) {
+      cluster_->statsScope().counter("ratelimit.failure_mode_allowed").inc();
+      if (!initiating_call_) {
+        callbacks_->continueDecoding();
+      }
+    } else {
+      state_ = State::Responded;
+      callbacks_->sendLocalReply(
+          ThriftProxy::AppException(ThriftProxy::AppExceptionType::InternalError, "limiter error"),
+          false);
+      callbacks_->streamInfo().setResponseFlag(StreamInfo::ResponseFlag::RateLimitServiceError);
+    }
+  } else if (!initiating_call_) {
```

Comment:
It doesn't. If `initiating_call_` is true here, it means that the completion callback occurred during the call to `client_->limit()` in `initiateCall`. By the time `initiateCall` returns, `state_` will be `Complete` and `messageBegin` will return `Continue`. Since the filter chain was  stopped, there's no need to call `continueDecoding`.

Reply:
`state_` may also be `Responded` but then no call to continueDecoding is needed because the rate limit filter short-circuited the request.